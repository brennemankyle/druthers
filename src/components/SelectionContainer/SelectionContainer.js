import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import AppPropTypes from '../../utils/AppPropTypes'

let SelectionContainer = (props) => {
  const searchRef = useRef(null)

  let onFocus = (e) => {
    if (!props.areOptionsOpen) {
      props.onFocus(e)

      searchRef.current.focus()
    } else {
      e.target.blur() // Close if already open options
    }
  }
  let onBlur = (e) => {
    if (e.target.classList.contains('search')) {
      setTimeout(() => { // Call blur after focus event
        props.onBlur(e)
      })
    }
  }
  let onSearchClick = (e) => {
    if (props.areOptionsOpen) {
      // Call blur on search click
      e.preventDefault()
      searchRef.current.blur()
      console.log('here')
    }
  }

  const Search = React.cloneElement(props.Search, {
    onClick: onSearchClick,
    ref: searchRef,
  })

  return <div
    className={props.className}
    tabIndex={props.styles.disabled ? 'false' : '-1'}
    onFocus={onFocus}
    onBlur={onBlur}>
      {props.SelectionList}
      {Search}
      <span className="divider"></span>
      <img className="expand" src="expand.png" alt="expand" /></div>
}

SelectionContainer.propTypes = {
  onFocus: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  areOptionsOpen: PropTypes.bool.isRequired,
  Search: AppPropTypes.element.isRequired,
  SelectionList: AppPropTypes.element.isRequired,
  styles: AppPropTypes.styles.isRequired,
}

export default SelectionContainer
