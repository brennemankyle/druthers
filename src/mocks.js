import React from 'react'

let noop = () => {}

let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

let mockEvent = {
  preventDefault: noop
}

let mockElement = (props) => <div data-props={props} />
let mockInput = (props) => <input data-props={props} />

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  mockElement,
  mockInput, }
