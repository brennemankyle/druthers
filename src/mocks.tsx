import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
