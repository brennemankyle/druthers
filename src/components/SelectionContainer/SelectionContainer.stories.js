import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { mockStyles } from '../../mocks'
import props from '../../utils/defaultProps'
import { SelectionContainer, SelectionList, Search, Selection } from '../styledComponents/styledComponents'

let selectionList = <SelectionList
  itemList={props.selection}
  onClick={action('onRemove')}
  removable={!props.disabled && props.removable}
  Item={Selection}
  styles={mockStyles} />
let search = <Search
  hide={false}
  placeholder={props.placeholder}
  searchText={''}
  onKeyDown={action('onKeyDown')}
  onChange={action('onChange')}
  styles={mockStyles} />

storiesOf('(Internal) SelectionContainer', module)
  .add('closed', () => <SelectionContainer
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    styles={mockStyles}
    areOptionsOpen={false}
    SelectionList={selectionList}
    Search={search} />)

storiesOf('(Internal) SelectionContainer', module)
  .add('open', () => <SelectionContainer
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    styles={mockStyles}
    areOptionsOpen={true}
    SelectionList={selectionList}
    Search={search} />)

let styles = {
  ...mockStyles,
  disabled: true,
}
storiesOf('(Internal) SelectionContainer', module)
  .add('disabled', () => <SelectionContainer
    onFocus={action('onFocus')}
    onBlur={action('onBlur')}
    styles={styles}
    areOptionsOpen={false}
    SelectionList={<SelectionList
      itemList={props.selection}
      onClick={action('onRemove')}
      removable={!props.disabled && props.removable}
      Item={Selection}
      styles={styles} />}
    Search={<Search
      placeholder={props.placeholder}
      searchText={''}
      onKeyDown={action('onKeyDown')}
      onChange={action('onChange')}
      styles={styles} />} />)
