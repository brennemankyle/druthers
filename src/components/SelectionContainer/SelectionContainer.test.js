import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { mockStyles, mockEvent, mockElement, mockInput } from '../../mocks'
import SelectionContainer from './SelectionContainer'

let onFocus = jest.fn()
let onBlur = jest.fn()

it('renders', () => {
  const wrapper = shallow(<SelectionContainer
    styles={mockStyles}
    className="test"
    areOptionsOpen={false}
    onFocus={onFocus}
    onBlur={onBlur}
    SelectionList={mockElement}
    Search={mockInput} />)

  expect(toJson(wrapper)).toMatchSnapshot()
})

it('renders disabled', () => {
  let styles = {
    disabled: true,
    ...mockStyles
  }

  const wrapper = shallow(<SelectionContainer
    styles={styles}
    className="test"
    areOptionsOpen={false}
    onFocus={onFocus}
    onBlur={onBlur}
    SelectionList={mockElement}
    Search={mockInput} />)

  expect(toJson(wrapper)).toMatchSnapshot()
})
