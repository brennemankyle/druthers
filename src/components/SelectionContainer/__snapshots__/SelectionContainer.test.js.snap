// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="test"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="-1"
>
  <Component />
  <Component
    onClick={[Function]}
  />
  <span
    className="divider"
  />
  <img
    alt="expand"
    className="expand"
    src="expand.png"
  />
</div>
`;

exports[`renders disabled 1`] = `
<div
  className="test"
  onBlur={[Function]}
  onFocus={[Function]}
  tabIndex="false"
>
  <Component />
  <Component
    onClick={[Function]}
  />
  <span
    className="divider"
  />
  <img
    alt="expand"
    className="expand"
    src="expand.png"
  />
</div>
`;
